src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp

```diff
@@ -265,45 +387,155 @@ struct boss_jeklik : public BossAI
+
+        switch (_mode)
         {
-            DoCastRandomTarget(SPELL_THROW_LIQUID_FIRE);
-            context.Repeat(7s);
-        });
+            case BAT_RIDER_MODE_BOSS:
+                events.ScheduleEvent(EVENT_BAT_RIDER_THROW_BOMB, 2s);
+                break;
+            case BAT_RIDER_MODE_TRASH:
+                events.ScheduleEvent(EVENT_DEMO_SHOUT, 1s);
+                events.ScheduleEvent(EVENT_BATTLE_COMMAND, 8s);
+                events.ScheduleEvent(EVENT_INFECTED_BITE, 6500ms);
+        }
     }
 
     void UpdateAI(uint32 diff) override
     {
-        if (!UpdateVictim())
-            return;
+        events.Update(diff);
 
-        _scheduler.Update(diff);
-    }
+        if (_mode == BAT_RIDER_MODE_BOSS)
+        {
+            // if the creature isn't moving, run the loop
+            if (!me->isMoving())
+            {
+                LOG_DEBUG("scripts.ai", "npc_batrider::UpdateAI: not moving, running loop");
+                // enable flying
+                me->SetDisableGravity(true);
+                // send the rider on its loop
+                me->GetMotionMaster()->MoveSplinePath(PATH_BAT_RIDER_LOOP);
+            }
+
+            // event handling
+            switch (events.ExecuteEvent())
+            {
+                case EVENT_BAT_RIDER_THROW_BOMB:
+                    LOG_DEBUG("scripts.ai", "npc_batrider::UpdateAI: EVENT_BAT_RIDER_THROW_BOMB");
+                    DoCastRandomTarget(SPELL_THROW_LIQUID_FIRE);
+                    events.ScheduleEvent(EVENT_BAT_RIDER_THROW_BOMB, 8s);
+                    break;
+                default:
+                    break;
+            }
+        }
+        else if (_mode == BAT_RIDER_MODE_TRASH)
+        {
+            if (!UpdateVictim())
+            {
+                return;
+            }
+
+            // don't interrupt casting
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+            {
+                return;
+            }
 
-private:
-    TaskScheduler _scheduler;
+            // if health goes below 30%, cast Unstable Concotion
+            if (me->HealthBelowPct(30))
+            {
+                events.ScheduleEvent(EVENT_UNSTABLE_CONCOCTION, 0);
+            }
+
+            // event handling
+            switch (events.ExecuteEvent())
+            {
+                case EVENT_DEMO_SHOUT:
+                    DoCast(me, SPELL_DEMO_SHOUT);
+                    break;
+                case EVENT_BATTLE_COMMAND:
+                    DoCast(me, SPELL_BATTLE_COMMAND);
+                    events.ScheduleEvent(EVENT_BATTLE_COMMAND, 25s);
+                    break;
+                case EVENT_INFECTED_BITE:
+                    DoCastVictim(SPELL_INFECTED_BITE);
+                    events.ScheduleEvent(EVENT_INFECTED_BITE, 8s);
+                    break;
+                case EVENT_UNSTABLE_CONCOCTION:
+                    Talk(EMOTE_LOW_HEALTH);
+                    DoCast(me, SPELL_UNSTABLE_CONCOCTION);
```

Comment:
```suggestion
                    DoCastSelf(SPELL_UNSTABLE_CONCOCTION);
```

and so on for the others as well

Reply:
Done.